searx/shared/redisdb.py

```diff
@@ -0,0 +1,41 @@
+# SPDX-License-Identifier: AGPL-3.0-or-later
+# lint: pylint
+"""Implementation of the redis client (redis-py_).
+
+.. _redis-py: https://github.com/redis/redis-py
+
+This implementation uses the :ref:`settings redis` setup from ``settings.yml``.
+A redis DB connect can be tested by::
+
+  >>> from searx.shared import redisdb
+  >>> redisdb.init()
+  True
+  >>> db = redisdb.client()
+  >>> db.set("foo", "bar")
+  True
+  >>> db.get("foo")
+  b'bar'
+  >>>
+
+"""
+
+import logging
+import redis
+from searx import get_setting
+
+logger = logging.getLogger('searx.shared.redis')
+
+
+def client():
+    return redis.Redis.from_url(get_setting('redis.url'))
```

Comment:
What about this:
```python
_client = None

def client():
    global _client
    if _client:
        return _client
    # not thread: safe: worst case scenario a two or more clients are initialized
    # only one is kept, the others are garbage collected.
    _client = redis.Redis.from_url(get_setting('redis.url'))
    return _client
```
?

So each call to the `client()` skip the overhead of the connection initialization.

AFIU, the client is [thread safe](https://github.com/redis/redis-py#thread-safety) and there is a [connection pool](https://github.com/redis/redis-py#connection-pools).

Reply:
> AFIU, the client is thread safe and there is a connection pool.

`redis.Redis.from_url` just wraps `ConnectionPool.from_url` ..

https://github.com/redis/redis-py/blob/15f315a496c3267c8cbcc6d6d9c6005ea4d4a4d5/redis/client.py#L852-L853

> So each call to the `client()` skip the overhead of the connection initialization.

AFAIK there is no need for, since `redis.Redis.from_url` just wraps `ConnectionPool.from_url` all the connection management and garbage collection is done by the ConnectionPool.

Reply:
@dalf: thanks for your hint / fixed in dca83944